pkg/services/baremetal/host/host.go

```diff
@@ -1789,65 +1804,222 @@ func analyzeSSHOutputProvisioned(out sshclient.Output) (isTimeout, isConnectionR
+			if err := handleSSHError(out); err != nil {
+				conditions.MarkFalse(host, infrav1.ProvisionSucceededCondition,
+					infrav1.StillProvisioningReason,
+					clusterv1.ConditionSeverityWarning, "%s",
+					err.Error())
+				return actionError{err: err}
 			}
-			if err != nil {
-				return actionError{err: fmt.Errorf(errMsgFailedHandlingIncompleteBoot, err)}
+		} else {
+			rebootType := infrav1.RebootTypeHardware
+			if _, err := s.scope.RobotClient.RebootBMServer(host.Spec.ServerID, rebootType); err != nil {
+				s.handleRobotRateLimitExceeded(err, rebootServerStr)
+				err = fmt.Errorf("actionProvisioned (Reboot via Annotation), reboot (%s) failed: %w", rebootType, err)
+				conditions.MarkFalse(host, infrav1.ProvisionSucceededCondition,
+					infrav1.StillProvisioningReason,
+					clusterv1.ConditionSeverityWarning, "%s",
+					err.Error())
+				return actionError{err: err}
 			}
-			return actionContinue{delay: 10 * time.Second}
-		}
-		// Reboot now
-		out := sshClient.Reboot()
-		if err := handleSSHError(out); err != nil {
-			return actionError{err: err}
 		}
+		msg := fmt.Sprintf("Rebooting because annotation was set. Old BootID: %s", currentBootID)
+		createSSHRebootEvent(ctx, host, msg)
+		conditions.MarkFalse(host, infrav1.ProvisionSucceededCondition,
+			infrav1.StillProvisioningReason,
+			clusterv1.ConditionSeverityInfo, "%s",
+			msg)
+		host.Spec.Status.Rebooted = true
+		return actionContinue{delay: 10 * time.Second}
+	}
+
+	// Reboot has been done already. Check whether it has been successful
+
+	if host.Spec.Status.ExternalIDs.RebootAnnotationNodeBootID == "" {
+		err := errors.New("internal error host.Spec.Status.ExternalIDs.RebootAnnotationNodeBootID not set")
+		s.scope.Logger.Error(err, "")
+		conditions.MarkFalse(host, infrav1.ProvisionSucceededCondition,
+			infrav1.StillProvisioningReason,
+			clusterv1.ConditionSeverityWarning, "%s",
+			err.Error())
+		return actionError{err: err}
+	}
 
-		createSSHRebootEvent(ctx, s.scope.HetznerBareMetalHost, "Rebooting because annotation was set")
-		s.scope.HetznerBareMetalHost.Spec.Status.Rebooted = true
+	if host.Spec.Status.ExternalIDs.RebootAnnotationNodeBootID != currentBootID {
+		// Reboot has been successful
+		s.scope.Logger.Info(fmt.Sprintf("BootID changed: %q -> %q", host.Spec.Status.ExternalIDs.RebootAnnotationNodeBootID, currentBootID))
+		host.Spec.Status.Rebooted = false
+		host.Spec.Status.ExternalIDs.RebootAnnotationNodeBootID = ""
+		host.Spec.Status.ExternalIDs.RebootAnnotationSince.Time = time.Time{}
+
+		host.ClearRebootAnnotations()
+
+		host.ClearError()
+		conditions.MarkTrue(host, infrav1.ProvisionSucceededCondition)
+		return actionComplete{}
+	}
+
+	s.scope.Logger.Info("foooo", "long", host.Spec.Status.ExternalIDs.RebootAnnotationNodeBootID,
+		"cur", currentBootID)
+
+	if !s.scope.SSHAfterInstallImage {
+		// s.scope.SSHAfterInstallImage is false: No ssh allowed.
+		// We can only wait for the BootID in the wl-cluster to change.
```

Comment:
We should include a timeout

Reply:
I added that timeout at the top of the func:

```go 
	rebootDuration := time.Since(host.Spec.Status.ExternalIDs.RebootAnnotationSince.Time)
	if rebootDuration > 5*time.Minute {
		msg := fmt.Sprintf("Rebooting timed out after: %s", rebootDuration.Round(time.Second))
		s.scope.Logger.Info(msg)
		conditions.MarkFalse(
			s.scope.HetznerBareMetalHost,
			infrav1.HostHasNoRebootAnnotationCondition,
			"TimedOut",
			clusterv1.ConditionSeverityError,
			"%s",
			msg,
		)
		return s.recordActionFailure(infrav1.FatalError, msg)
	}
```